File: java/core/src/main/java/network/aika/typedefs/LinkDefinition.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package network.aika.typedefs;

import network.aika.activations.Activation;
import network.aika.activations.Link;
import network.aika.type.relations.Relation;
import network.aika.type.relations.RelationOne;
import network.aika.neurons.Synapse;
import network.aika.type.Type;
import network.aika.type.TypeRegistry;
import network.aika.type.relations.RelationSelf;

import java.util.List;
import java.util.Set;
import java.util.stream.Stream;

/**
 *
 * @author Lukas Molzberger
 */
public class LinkDefinition extends Type {

    public static final RelationSelf SELF = new RelationSelf(0, "LINK-SELF");

    public static final RelationOne INPUT = new RelationOne(1, "LINK-INPUT");
    public static final RelationOne OUTPUT = new RelationOne(2, "LINK-OUTPUT");
    public static final RelationOne SYNAPSE = new RelationOne(3, "LINK-SYNAPSE");

    public static final RelationOne CORRESPONDING_INPUT_LINK = new RelationOne(4, "CORRESPONDING_INPUT_LINK");
    public static final RelationOne CORRESPONDING_OUTPUT_LINK = new RelationOne(5, "CORRESPONDING_OUTPUT_LINK");

    public static final Relation[] RELATIONS = {SELF, INPUT, OUTPUT, SYNAPSE, CORRESPONDING_INPUT_LINK, CORRESPONDING_OUTPUT_LINK};

    static {
        SYNAPSE.setReversed(SynapseDefinition.LINK);
        INPUT.setReversed(ActivationDefinition.OUTPUT);
        OUTPUT.setReversed(ActivationDefinition.INPUT);
        CORRESPONDING_INPUT_LINK.setReversed(CORRESPONDING_OUTPUT_LINK);
        CORRESPONDING_OUTPUT_LINK.setReversed(CORRESPONDING_INPUT_LINK);
    }

    private SynapseDefinition synapse;

    private ActivationDefinition input;
    private ActivationDefinition output;


    public LinkDefinition(TypeRegistry registry, String name) {
        super(registry, name);
    }

    @Override
    public Relation[] getRelations() {
        return RELATIONS;
    }

    public Link instantiate(Synapse synapse, Activation input, Activation output) {
        assert input.getType().isInstanceOf(getInput());
        assert output.getType().isInstanceOf(getOutput());

        return new Link(this, synapse, input, output);
    }

    public SynapseDefinition getSynapse() {
        return synapse != null ?
                synapse :
                getFromParent(p -> ((LinkDefinition)p).getSynapse());
    }

    public LinkDefinition setSynapse(SynapseDefinition synapse) {
        assert synapse != null;

        this.synapse = synapse;

        return this;
    }

    public ActivationDefinition getInput() {
        return input != null ?
                input :
                getFromParent(p -> ((LinkDefinition)p).getInput());
    }

    public LinkDefinition setInput(ActivationDefinition input) {
        assert input != null;

        this.input = input;

        return this;
    }

    public ActivationDefinition getOutput() {
        return output != null ?
                output :
                getFromParent(p -> ((LinkDefinition)p).getOutput());
    }

    public LinkDefinition setOutput(ActivationDefinition output) {
        assert output != null;

        this.output = output;
        return this;
    }

    @Override
    public String toString() {
        return super.toString() + " (" + getInput().getName() + " --> " + getOutput().getName() + ")";
    }
}

File: src/network/link_definition.cpp
#include "network/link_definition.h"

const RelationSelf LinkDefinition::SELF = RelationSelf(0, "SELF");
const RelationOne LinkDefinition::INPUT = RelationOne(1, "INPUT");
const RelationOne LinkDefinition::OUTPUT = RelationOne(2, "OUTPUT");
const RelationOne LinkDefinition::SYNAPSE = RelationOne(3, "SYNAPSE");
const RelationOne LinkDefinition::PAIR_IN = RelationOne(4, "PAIR_IN");
const RelationOne LinkDefinition::PAIR_OUT = RelationOne(5, "PAIR_OUT");

// Static initializer to set up reverse relationships
class LinkDefinitionInitializer {
public:
    LinkDefinitionInitializer() {
        // Set up bidirectional relationships
        const_cast<RelationOne&>(LinkDefinition::INPUT).setReversed(const_cast<RelationOne*>(&LinkDefinition::OUTPUT));
        const_cast<RelationOne&>(LinkDefinition::OUTPUT).setReversed(const_cast<RelationOne*>(&LinkDefinition::INPUT));
        const_cast<RelationOne&>(LinkDefinition::PAIR_IN).setReversed(const_cast<RelationOne*>(&LinkDefinition::PAIR_OUT));
        const_cast<RelationOne&>(LinkDefinition::PAIR_OUT).setReversed(const_cast<RelationOne*>(&LinkDefinition::PAIR_IN));
        
        // SYNAPSE and SELF are their own reverse
        const_cast<RelationOne&>(LinkDefinition::SYNAPSE).setReversed(const_cast<RelationOne*>(&LinkDefinition::SYNAPSE));
        const_cast<RelationSelf&>(LinkDefinition::SELF).setReversed(const_cast<RelationSelf*>(&LinkDefinition::SELF));
    }
};

static LinkDefinitionInitializer linkDefInit;

// Can't use an abstract class Relation in a vector directly
// Will need to implement a different approach
// Commenting out for now to fix compilation
// const std::vector<Relation> LinkDefinition::RELATIONS = {SELF, INPUT, OUTPUT, SYNAPSE, PAIR_IN, PAIR_OUT};

LinkDefinition::LinkDefinition(TypeRegistry* registry, const std::string& name) : Type(registry, name) {}

std::vector<Relation*> LinkDefinition::getRelations() const {
    // Return a vector of pointers to avoid the abstract class issue
    return {
        const_cast<RelationSelf*>(&SELF),
        const_cast<RelationOne*>(&INPUT),
        const_cast<RelationOne*>(&OUTPUT),
        const_cast<RelationOne*>(&SYNAPSE),
        const_cast<RelationOne*>(&PAIR_IN),
        const_cast<RelationOne*>(&PAIR_OUT)
    };
}

Link* LinkDefinition::instantiate(Synapse* synapse, Activation* input, Activation* output) {
    // Implementation for instantiating a Link
    return nullptr; // Placeholder
}

SynapseDefinition* LinkDefinition::getSynapse() const {
    return synapse;
}

LinkDefinition* LinkDefinition::setSynapse(SynapseDefinition* synapse) {
    this->synapse = synapse;
    return this;
}

ActivationType* LinkDefinition::getInput() const {
    return input;
}

LinkDefinition* LinkDefinition::setInput(ActivationType* input) {
    this->input = input;
    return this;
}

ActivationType* LinkDefinition::getOutput() const {
    return output;
}

LinkDefinition* LinkDefinition::setOutput(ActivationType* output) {
    this->output = output;
    return this;
}

std::string LinkDefinition::toString() const {
    return "LinkDefinition: " + name;
} 